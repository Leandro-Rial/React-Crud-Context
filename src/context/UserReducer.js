import { types } from "../types/types"

export const UserReducer = (state = {}, action) => {
    switch (action.type) {
        case types.removeUser:
            
            return {
                ...state,
                users: state.users.filter(user => {
                    return user.id !== action.payload
                })
            }
        
        case types.addUser:
            
            return {
                ...state,
                users: [action.payload, ...state.users]
            }
        
        
        case types.editUser:
            const updateUser = action.payload;
            
            const updateUsers = state.users.map(user => {
                if(user.id === updateUser.id) {
                    return updateUser
                }

                return user;
            })

            return {
                ...state,
                users: updateUsers
            }
        
    
        default:
            return state
    }
}