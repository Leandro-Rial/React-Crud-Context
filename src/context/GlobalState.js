import React, { createContext, useEffect, useReducer } from 'react';
import { types } from '../types/types';
import { UserReducer } from './UserReducer';

const initialState = {
    users: []
}

const init = () => {
    return (
        JSON.parse(localStorage.getItem("users")) || initialState
    )
}

export const UserContext = createContext(initialState);

export const UserProvider = ({ children }) => {

    const [state, dispatch] = useReducer(UserReducer, initialState, init)

    useEffect(() => {
        localStorage.setItem('users', JSON.stringify(state))
    }, [state])

    const removeUser = (id) => {
        dispatch({
            type: types.removeUser,
            payload: id
        })
    }
    
    const addUser = (user) => {
        dispatch({
            type: types.addUser,
            payload: user,
        })
    }

    const editUser = (user) => {
        dispatch({
            type: types.editUser,
            payload: user
        })
    }

    return (
        <UserContext.Provider value={{
            users: state.users,
            removeUser,
            addUser,
            editUser
        }}>
            {children}
        </UserContext.Provider>
    )
    
}