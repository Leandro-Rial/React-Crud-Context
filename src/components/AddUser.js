import React, { useContext, useState } from 'react';
import { UserContext } from '../context/GlobalState';
import { Link, useHistory } from 'react-router-dom';
import {
    Form,
    FormGroup,
    Label,
    Input,
    Button
} from 'reactstrap'
import { v4 as uuid } from 'uuid'

const AddUser = () => {

    const [name, setName] = useState('');

    const { addUser } = useContext(UserContext);

    const history = useHistory()

    const handleInputChange = (e) => {
        setName(e.target.value)
    }

    const handleSubmitForm = (e) => {
        e.preventDefault();

        const newUser = {
            id: uuid(),
            name
        }

        addUser(newUser)

        history.push('/')
    }

    return (
        <Form onSubmit={handleSubmitForm}>
            <FormGroup>
                <Label>Name</Label>
                <Input type="text" placeholder="Enter name" name="name" value={name} onChange={handleInputChange} required></Input>
            </FormGroup>
            <Button type="submit">Submit</Button>
            <Link to="/" className="btn btn-danger ml-2">Cancel</Link>
        </Form>
    )
}

export default AddUser
