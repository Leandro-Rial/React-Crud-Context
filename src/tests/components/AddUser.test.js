import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import AddUser from "../../components/AddUser";
import { UserProvider } from "../../context/GlobalState";

describe('Pruebas en AddUser', () => {
    
    const wrapper = mount(
        <UserProvider>
            <MemoryRouter>
                <AddUser />
            </MemoryRouter>
        </UserProvider>
    )
    
    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    });

    test('debe de realizar el submit', () => {
        wrapper.find("form").prop(onsubmit, {
            preventDefault(){}
        })
    });
});