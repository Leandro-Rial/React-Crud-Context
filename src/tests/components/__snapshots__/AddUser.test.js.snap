// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddUser debe de mostrarse correctamente 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Name
    </label>
    <input
      className="form-control"
      name="name"
      onChange={[Function]}
      placeholder="Enter name"
      required={true}
      type="text"
      value=""
    />
  </div>
  <button
    aria-label={null}
    className="btn btn-secondary"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
  <a
    className="btn btn-danger ml-2"
    href="/"
    onClick={[Function]}
  >
    Cancel
  </a>
</form>
`;
